homeassistant/components/xiaomi_miio/__init__.py

```diff
@@ -1 +1,79 @@
 """Support for Xiaomi Miio."""
+import logging
+
+from miio import DeviceException, gateway
+
+from homeassistant import config_entries, core
+from homeassistant.const import CONF_HOST, CONF_TOKEN
+from homeassistant.helpers import device_registry as dr
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+GATEWAY_PLATFORMS = ["alarm_control_panel"]
+
+
+async def async_setup(hass: core.HomeAssistant, config: dict):
+    """Set up the Xiaomi Miio component."""
+    conf = config.get(DOMAIN)
+    hass.data.setdefault(DOMAIN, {})
+
+    if not conf:
+        return True
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": config_entries.SOURCE_IMPORT}, data=conf
+        )
+    )
+    return True
+
+
+async def async_setup_entry(
+    hass: core.HomeAssistant, entry: config_entries.ConfigEntry
+):
+    """Set up the Xiaomi Miio component from a config entry."""
+    host = entry.data.get(CONF_HOST)
+    token = entry.data.get(CONF_TOKEN)
+    name = entry.title
+    gateway_id = entry.data.get("gateway_id")
+    _LOGGER.info("Initializing with host %s (token %s...)", host, token[:5])
+
+    # Connect to gateway
+    try:
+        gateway_device = gateway.Gateway(host, token)
+        gateway_info = gateway_device.info()
```

Comment:
What do you mean with I/O?
This is the place where the gateway_device class is setup, that class connects to the device during its initilization and that class contains all functions to communicate (get/set) with the device.

Reply:
The gateway_info is only a class that contains information, it does not get or set anything, it is only needed to get the model number, firmware version etc.

Reply:
Input/Output. Eg making a request over the network, reading or writing to disk etc.

Reply:
so the `gateway_device = gateway.Gateway(host, token)` is I/O since it makes the connection to the device.
The `gateway_info = gateway_device.info()` is also I/O since it makes a request to get the gateway info

Reply:
Ok, then we need to schedule these calls on the executor thread pool with `hass.async_add_executor_job`.

Try to group both calls into one function that stores or returns the variables. Maybe consider wrapping this into another class that has a setup function that sets the gateway and info as attributes.

Reply:
I split up async_setup_entry to allow for easy addition of the other Xiaomi Miio devices such as robot vacuums.
I also centrilized the connect logic in the gateway.py file in the ConnectXiaomiGateway class.
And I made this connection async to not block hass.